core/psh/nc.c

```diff
@@ -0,0 +1,337 @@
+				}
+
+				return sfd;
+			}
+			else {
+				return -1;
+			}
+		}
+	}
+}
+
+static void psh_nc_socktalk(int fd)
+{
+	fd_set fds;
+	int n;
+	char buf[64];
+
+	while (1) {
+		FD_ZERO(&fds);
+		FD_SET(STDIN_FILENO, &fds);
+		FD_SET(fd, &fds);
+
+		if (select(fd + 1, &fds, NULL, NULL, NULL) <= 0) {
+			return;
+		}
+
+		if (FD_ISSET(STDIN_FILENO, &fds)) {
+			if ((n = read(STDIN_FILENO, buf, sizeof(buf))) <= 0)
+				return;
+
+			if (send(fd, buf, n, 0) <= 0)
+				return;
+			memset(buf, 0, n);
+		}
+
+		if (FD_ISSET(fd, &fds)) {
+			if ((n = recv(fd, buf, sizeof(buf), 0)) <= 0)
+				return;
+			write(STDOUT_FILENO, buf, n);
+			memset(buf, 0, n);
+		}
+	}
+}
+
+int psh_nc(int argc, char *argv[])
+{
+	struct sockaddr_storage srcsockaddr, dstsockaddr;
+	int c, l = 0;
+	int socktype = SOCK_STREAM;
+	char *srcaddr = NULL, *srcport = NULL, *dstaddr = NULL, *dstport = NULL;
+	int af_family = AF_UNSPEC;
+	int fd, cfd;
+	socklen_t addrlen;
+
+	while ((c = getopt(argc, argv, "lu46s:p:")) != -1) {
+		switch (c) {
+		case 'l':
+			l = 1;
+			break;
+		case 'u':
+			socktype = SOCK_DGRAM;
+			break;
+		case '6':
+			af_family = AF_INET6;
+			break;
+		case '4':
+			af_family = AF_INET;
+			break;
+		case 's':
+			srcaddr = optarg;
+			break;
+		case 'p':
+			srcport = optarg;
+			break;
+		default:
+			fprintf(stderr, "nc: Unknown option!\n");
+			return -EINVAL;
+		}
+	}
+
+	argc -= optind;
+	argv += optind;
+
+	if (l) {
+		switch (argc) {
+		case 2:
+			if (srcaddr || srcport) {
+				fprintf(stderr, "nc: Too many arguments!\n");
+				return -EINVAL;
+			}
+			srcaddr = *argv;
+			srcport = *(argv + 1);
+			break;
+		case 1:
+			if (!srcport) {
+				srcport = *argv;
+			} else if (!srcaddr) {
+				srcaddr = *argv;
+			} else {
+				fprintf(stderr, "nc: Too many arguments!\n");
+				return -EINVAL;
+			}
+			break;
+		default:
+			fprintf(stderr, "nc: Unexpected number of arguments!\n");
+			return -EINVAL;
+		}
+    }
+	else {
+		if (argc < 2) {
+			fprintf(stderr, "nc: Expected a port number!\n");
+			return -EINVAL;
+		}
+
+		if (argc < 1) {
+			fprintf(stderr, "nc: Expected an address and a port number!\n");
+			return -EINVAL;
+		}
+		dstaddr = *argv;
+		dstport = *(argv + 1);
+
+		if (psh_nc_sockaddrFill(&af_family, dstaddr, dstport,
+		                        &dstsockaddr, &addrlen) < 0)
+			return -EINVAL;
+	}
+
+	/* Optional for client and required for server */
+	if (srcport || srcaddr) {
+		if (psh_nc_sockaddrFill(&af_family, srcaddr, srcport,
+		                        &srcsockaddr, &addrlen) < 0)
+			return -EINVAL;
+	}
+
+	if ((fd = socket(af_family, socktype, socktype == SOCK_STREAM ?
+	                                      IPPROTO_TCP : IPPROTO_UDP)) < 0) {
+		fprintf(stderr, "nc: Can't create a socket!\n");
+		return -EIO;
+	}
+
+	if (srcport || srcaddr) {
+		if (bind(fd, (struct sockaddr *) &srcsockaddr, addrlen) < 0) {
+			fprintf(stderr, "nc: Can't bind to a socket!\n");
+			close(fd);
+			return -EIO;
+		}
+	}
+
+	if (l) {
```

Comment:
What is this `if (1) `{ for?
`#if 0` is preffered form for dead code storage ;)

Reply:
> Kinda too much line breaking in this code, makes is less readable. I think that 120 chars wide line is ok limit, we have now huge 16:9 or 21:9 LCDs, so we can afford that.

I agree, I was trying to follow the classic convention of using up to 80 character lines.